packages/compass-components/.eslintrc.js

```diff
@@ -0,0 +1,19 @@
+module.exports = {
+  root: true,
+  parser: '@typescript-eslint/parser',
+  plugins: [
+    '@typescript-eslint',
+    'jsx-a11y'
+  ],
+  extends: [
+    'eslint:recommended',
+    'plugin:@typescript-eslint/recommended'
+  ],
```

Comment:
this is quite loose, would it be possible to make it consistent with the other products? in mongosh (https://github.com/mongodb-js/mongosh/blob/main/config/eslintrc.base.js) we also lint js somehow and we kept the style consistent with `eslint-config-mongodb-js` which will help a bit with the conversion and to keep things more aligned.

Also these overrides there makes lot of sense if they are not pulled in by default:

```
    '@typescript-eslint/no-unused-vars': 2,
    '@typescript-eslint/no-floating-promises': 2,
    '@typescript-eslint/await-thenable': 2,
    '@typescript-eslint/require-await': 2,
    '@typescript-eslint/explicit-module-boundary-types': 0,
    '@typescript-eslint/ban-types': 0,
    'mocha/no-skipped-tests': 1,
    'mocha/no-exclusive-tests': 2,
```

Reply:
They’re not pulled in by default in mongosh, yup

Reply:
`+1` on this one, I'd also prefer having it quite tight, especially also including linting rules for the tests, like `no-skipped` / `no-exclusive` things - they saved my butt a couple times already 👍

Reply:
I have been burned by poorly maintained custom configuration files for code linting and styling too much in the past, I _really_ want us to spend as little time as possible looking at this config in the future (without them being abandoned and getting outdated like `mongodb-js-eslint`), so I have a strong preference of maintaining as little of custom configurations as possible and would really like us not to include `mongodb-js-eslint` in any way in this new config (especially as we are planning to use it as a base for other packages in the monorepo) and in general would really love us to rely more on recommended configs and minimize the amount of rule overrides that we do in our configuration.

I totally get the concern that the recommended options might look quite loose especially as the first thing that you see in the configuration is just a few lines that extend from the recommended options, so I did a little comparison. I pulled a list of all rules that are currently used in mongosh + the ones suggested in https://github.com/mongodb-js/compass/pull/2368/files#r677604520 and diffed them with all the rules that eslint + typescript-eslint + react recommended would give us. At the first glance the difference is pretty big, mongosh config has 80+ rules that are missing from recommended ones, but from those 80:

- 28 are just style. FWIW eslint is not the best tool to maintain code style, but as I feel the general preference is to keep it like that, maybe we can take all those style rules and make _them_ into a separate shareable configuration and nothing else
- 6 are deprecated and should be replaced with other rules
- 2 are just redundant, don't make sense in combination with other rules. E.g., we have a `block-scoped-var` rule and a `no-var` rule at the same time
- 1 is literally a no-op because of how it is configured, lol

So we are left with ~40 rules out of ~140 that we probably should keep by just adding them directly to rules in this configuration (while maybe giving them another closer look, some of them don't look very useful to me tbh, things like `react/no-multi-comp` should definitely go away).

On top of that recommended rules have ~40 rules (none of them are style related 😉) that are not part of our current config, so all-in-all the recommended options give us more or less the same amount of checks, just different ones in my opinion. And I don't think it's a bad idea to incorporate those additional rules in our configuration anyway.

Here's an overview of all the rules that are part of our current configuration in mongosh:

❌ - redundant
🏚 - deprecated
👀 - totally let's keep them, but give them another thorough look
💅 - code style rules

||Type|Description|
|---|---|---|
|❌|[block-scoped-var](https://eslint.org/docs/rules/block-scoped-var)|Treat var as Block Scoped (block-scoped-var)|
|❌|[no-mixed-requires](https://eslint.org/docs/rules/no-mixed-requires)|disallow require calls to be mixed with regular variable declarations (no-mixed-requires)|
|🏚|[callback-return](https://eslint.org/docs/rules/callback-return)|Enforce Return After Callback (callback-return)|
|🏚|[handle-callback-err](https://eslint.org/docs/rules/handle-callback-err)|Enforce Callback Error Handling (handle-callback-err)|
|🏚|[no-negated-in-lhs](https://eslint.org/docs/rules/no-negated-in-lhs)|disallow negating the left operand in in expressions (no-negated-in-lhs)|
|🏚|[no-new-require](https://eslint.org/docs/rules/no-new-require)|Disallow new require (no-new-require)|
|🏚|[no-path-concat](https://eslint.org/docs/rules/no-path-concat)|Disallow string concatenation when using __dirname and __filename (no-path-concat)|
|🏚|[no-sync](https://eslint.org/docs/rules/no-sync)|Disallow Synchronous Methods (no-sync)|
|👀|[@typescript-eslint/await-thenable](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/await-thenable.md)|Disallows awaiting a value that is not a Thenable (await-thenable)|
|👀|[@typescript-eslint/no-floating-promises](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md)|Requires Promise-like values to be handled appropriately (no-floating-promises)|
|👀|[@typescript-eslint/require-await](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/require-await.md)|Disallow async functions which have no await expression (require-await)|
|👀|[complexity](https://eslint.org/docs/rules/complexity)|Limit Cyclomatic Complexity (complexity)|
|👀|[default-case](https://eslint.org/docs/rules/default-case)|Require Default Case in Switch Statements (default-case)|
|👀|[dot-notation](https://eslint.org/docs/rules/dot-notation)|Require Dot Notation (dot-notation)|
|👀|[eqeqeq](https://eslint.org/docs/rules/eqeqeq)|Require === and !== (eqeqeq)|
|👀|[guard-for-in](https://eslint.org/docs/rules/guard-for-in)|Require Guarding for-in (guard-for-in)|
|👀|[id-length](https://eslint.org/docs/rules/id-length)|enforce minimum and maximum identifier lengths (id-length)|
|👀|[new-cap](https://eslint.org/docs/rules/new-cap)|require constructor names to begin with a capital letter (new-cap)|
|👀|[no-alert](https://eslint.org/docs/rules/no-alert)|Disallow Use of Alert (no-alert)|
|👀|[no-caller](https://eslint.org/docs/rules/no-caller)|Disallow Use of caller/callee (no-caller)|
|👀|[no-console](https://eslint.org/docs/rules/no-console)|disallow the use of console (no-console)|
|👀|[no-else-return](https://eslint.org/docs/rules/no-else-return)|Disallow return before else (no-else-return)|
|👀|[no-eval](https://eslint.org/docs/rules/no-eval)|Disallow eval() (no-eval)|
|👀|[no-extend-native](https://eslint.org/docs/rules/no-extend-native)|Disallow Extending of Native Objects (no-extend-native)|
|👀|[no-extra-bind](https://eslint.org/docs/rules/no-extra-bind)|Disallow unnecessary function binding (no-extra-bind)|
|👀|[no-implied-eval](https://eslint.org/docs/rules/no-implied-eval)|Disallow Implied eval() (no-implied-eval)|
|👀|[no-iterator](https://eslint.org/docs/rules/no-iterator)|Disallow Iterator (no-iterator)|
|👀|[no-labels](https://eslint.org/docs/rules/no-labels)|Disallow Labeled Statements (no-labels)|
|👀|[no-lone-blocks](https://eslint.org/docs/rules/no-lone-blocks)|Disallow Unnecessary Nested Blocks (no-lone-blocks)|
|👀|[no-lonely-if](https://eslint.org/docs/rules/no-lonely-if)|disallow if statements as the only statement in else blocks (no-lonely-if)|
|👀|[no-loop-func](https://eslint.org/docs/rules/no-loop-func)|Disallow Functions in Loops (no-loop-func)|
|👀|[no-multi-str](https://eslint.org/docs/rules/no-multi-str)|Disallow Multiline Strings (no-multi-str)|
|👀|[no-native-reassign](https://eslint.org/docs/rules/no-native-reassign)|Disallow Reassignment of Native Objects (no-native-reassign)|
|👀|[no-nested-ternary](https://eslint.org/docs/rules/no-nested-ternary)|disallow nested ternary expressions (no-nested-ternary)|
|👀|[no-new](https://eslint.org/docs/rules/no-new)|Disallow new For Side Effects (no-new)|
|👀|[no-new-func](https://eslint.org/docs/rules/no-new-func)|Disallow Function Constructor (no-new-func)|
|👀|[no-new-object](https://eslint.org/docs/rules/no-new-object)|disallow Object constructors (no-new-object)|
|👀|[no-new-wrappers](https://eslint.org/docs/rules/no-new-wrappers)|Disallow Primitive Wrapper Instances (no-new-wrappers)|
|👀|[no-octal-escape](https://eslint.org/docs/rules/no-octal-escape)|disallow octal escape sequences in string literals (no-octal-escape)|
|👀|[no-proto](https://eslint.org/docs/rules/no-proto)|Disallow Use of __proto__ (no-proto)|
|👀|[no-return-assign](https://eslint.org/docs/rules/no-return-assign)|Disallow Assignment in return Statement (no-return-assign)|
|👀|[no-script-url](https://eslint.org/docs/rules/no-script-url)|Disallow Script URLs (no-script-url)|
|👀|[no-self-compare](https://eslint.org/docs/rules/no-self-compare)|Disallow Self Compare (no-self-compare)|
|👀|[no-sequences](https://eslint.org/docs/rules/no-sequences)|Disallow Use of the Comma Operator (no-sequences)|
|👀|[no-throw-literal](https://eslint.org/docs/rules/no-throw-literal)|Restrict what can be thrown as an exception (no-throw-literal)|
|👀|[no-unused-expressions](https://eslint.org/docs/rules/no-unused-expressions)|Disallow Unused Expressions (no-unused-expressions)|
|👀|[no-var](https://eslint.org/docs/rules/no-var)|require let or const instead of var (no-var)|
|👀|[one-var](https://eslint.org/docs/rules/one-var)|enforce variables to be declared either together or separately in functions (one-var)|
|👀|[prefer-const](https://eslint.org/docs/rules/prefer-const)|Suggest using const (prefer-const)|
|👀|[radix](https://eslint.org/docs/rules/radix)|Require Radix Parameter (radix)|
|👀|[react/jsx-boolean-value](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-boolean-value.md)|Enforce boolean attributes notation in JSX|
|👀|[react/no-did-mount-set-state](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-did-mount-set-state.md)|Prevent usage of setState in componentDidMount|
|👀|[react/no-did-update-set-state](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-did-update-set-state.md)|Prevent usage of setState in componentDidUpdate (react/no-did-update-set-state)|
|👀|[react/no-multi-comp](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-multi-comp.md)|Prevent multiple component definition per file (react/no-multi-comp)|
|👀|[react/self-closing-comp](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/self-closing-comp.md)|Prevent extra closing tags for components without children (react/self-closing-comp)|
|👀|[strict](https://eslint.org/docs/rules/strict)|require or disallow strict mode directives (strict)|
|👀|[yoda](https://eslint.org/docs/rules/yoda)|Require or disallow Yoda Conditions (yoda)|
|💅|[@typescript-eslint/semi](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/semi.md)|Require or disallow semicolons instead of ASI (semi)|
|💅|[brace-style](https://eslint.org/docs/rules/brace-style)|Require Brace Style (brace-style)|
|💅|[camelcase](https://eslint.org/docs/rules/camelcase)|Require CamelCase (camelcase)|
|💅|[comma-dangle](https://eslint.org/docs/rules/comma-dangle)|require or disallow trailing commas (comma-dangle)|
|💅|[comma-spacing](https://eslint.org/docs/rules/comma-spacing)|Enforces spacing around commas (comma-spacing)|
|💅|[comma-style](https://eslint.org/docs/rules/comma-style)|Comma style (comma-style)|
|💅|[curly](https://eslint.org/docs/rules/curly)|Require Following Curly Brace Conventions (curly)|
|💅|[eol-last](https://eslint.org/docs/rules/eol-last)|require or disallow newline at the end of files (eol-last)|
|💅|[indent](https://eslint.org/docs/rules/indent)|enforce consistent indentation (indent)|
|💅|[jsx-quotes](https://eslint.org/docs/rules/jsx-quotes)|enforce the consistent use of either double or single quotes in JSX attributes (jsx-quotes)|
|💅|[key-spacing](https://eslint.org/docs/rules/key-spacing)|enforce consistent spacing between keys and values in object literal properties (key-spacing)|
|💅|[keyword-spacing](https://eslint.org/docs/rules/keyword-spacing)|enforce consistent spacing before and after keywords (keyword-spacing)|
|💅|[no-extra-parens](https://eslint.org/docs/rules/no-extra-parens)|disallow unnecessary parentheses (no-extra-parens)|
|💅|[no-floating-decimal](https://eslint.org/docs/rules/no-floating-decimal)|Disallow Floating Decimals (no-floating-decimal)|
|💅|[no-multi-spaces](https://eslint.org/docs/rules/no-multi-spaces)|Disallow multiple spaces (no-multi-spaces)|
|💅|[no-multiple-empty-lines](https://eslint.org/docs/rules/no-multiple-empty-lines)|disallow multiple empty lines (no-multiple-empty-lines)|
|💅|[no-spaced-func](https://eslint.org/docs/rules/no-spaced-func)|disallow spacing between function identifiers and their applications (no-spaced-func)|
|💅|[no-trailing-spaces](https://eslint.org/docs/rules/no-trailing-spaces)|disallow trailing whitespace at the end of lines (no-trailing-spaces)|
|💅|[object-curly-spacing](https://eslint.org/docs/rules/object-curly-spacing)|enforce consistent spacing inside braces (object-curly-spacing)|
|💅|[padded-blocks](https://eslint.org/docs/rules/padded-blocks)|require or disallow padding within blocks (padded-blocks)|
|💅|[quotes](https://eslint.org/docs/rules/quotes)|enforce the consistent use of either backticks, double, or single quotes (quotes)|
|💅|[react/jsx-wrap-multilines](https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/jsx-wrap-multilines.md)|Prevent missing parentheses around multiline JSX (react/jsx-wrap-multilines)|
|💅|[semi-spacing](https://eslint.org/docs/rules/semi-spacing)|Enforce spacing before and after semicolons (semi-spacing)|
|💅|[space-before-blocks](https://eslint.org/docs/rules/space-before-blocks)|Require Or Disallow Space Before Blocks (space-before-blocks)|
|💅|[space-before-function-paren](https://eslint.org/docs/rules/space-before-function-paren)|Require or disallow a space before function parenthesis (space-before-function-paren)|
|💅|[space-infix-ops](https://eslint.org/docs/rules/space-infix-ops)|require spacing around infix operators (space-infix-ops)|
|💅|[spaced-comment](https://eslint.org/docs/rules/spaced-comment)|Requires or disallows a whitespace (space or tab) beginning a comment (spaced-comment)|
|💅|[wrap-iife](https://eslint.org/docs/rules/wrap-iife)|Require IIFEs to be Wrapped (wrap-iife)|

Reply:
Thanks for this, is super detailed @gribnoysup!

> So we are left with ~40 rules out of ~140 that we probably should keep by just adding them directly to rules in this configuration 

I think some of the one you have listed there may actually even be in the defaults .. maybe they are overridden in the `eslint-config-mongodb-js`? I'm sure stuff like `complexity` and  `quotes` comes for free.

I don't think that we have to use the `eslint-config-mongodb-js` unless that helps in any way. The defaults here in the PR are lacking few things though that are important practically speaking:
- don't forget await
- don't forget it.only on tests
- some rules for dependencies (some stuff are not detected by depcheck it seems), i think is some kind of recommended here: https://www.npmjs.com/package/eslint-plugin-node

Also I would love if we lint any js file (js and ts) also in the root / outside of plugins.

Reply:
> I'm sure stuff like `complexity` and `quotes` comes for free.

No, they are not part of recommended, `quotes` are style related and eslint doesn't put anything from that group there and for `complexity` I don't know for sure, but I'd guess that as the number in configuration is debatable it would be hard to put in recommended in a way that would be equally helpful for everyone (and recommended options usually try to avoid this kind of situation)

Reply:
Updated the lint rules, now we're using recommended rules from most of the plugins. Looking through the mentioned rules it seems to grab all of them. `plugin:@typescript-eslint/recommended`, `plugin:mocha/recommended`, etc